src/tail/session.rs

```diff
@@ -36,41 +48,96 @@ impl Session {
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+                let duration = Duration::from_millis(1000 * KEEP_ALIVE_INTERVAL);
+                let mut delay = delay_for(duration);
+
                 loop {
-                    thread::sleep(Duration::from_secs(60));
-                    let heartbeat_result = send_heartbeat(target, &client, &tail_id).await;
-                    if heartbeat_result.is_err() {
-                        return heartbeat_result;
+                    match rx.try_recv() {
+                        Err(TryRecvError::Empty) => {
+                            if delay.is_elapsed() {
+                                let heartbeat_result =
+                                    send_heartbeat(&target, &client, &tail_id).await;
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
                     }
-                    // This should loop forever until SIGINT is issued or Wrangler process is killed
-                    // through other means.
                 }
             }
             Err(e) => failure::bail!(http::format_error(e, None)),
         }
     }
 }
 
-async fn get_tunnel_url() -> Result<String, failure::Error> {
+async fn get_tunnel_url(rx: &mut Receiver<()>) -> Result<String, failure::Error> {
     // regex for extracting url from cloudflared metrics port.
     let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
 
-    let mut attempt = 0;
+    struct RetryDelay {
+        delay: Delay,
+        attempt: u64,
+        max_attempts: u64,
+    }
 
-    // This exponential backoff retry loop retries retrieving the cloudflared endpoint url
-    // from the cloudflared /metrics endpoint until it gets the URL or has tried retrieving the URL
-    // over 5 times.
-    while attempt < 5 {
-        if let Ok(resp) = reqwest::get("http://localhost:8081/metrics").await {
-            let body = resp.text().await?;
+    // This retry loop retries retrieving the cloudflared endpoint url from the cloudflared /metrics
+    // until it gets the URL or has tried retrieving the URL over 5 times.
+    impl RetryDelay {
+        fn new(max_attempts: u64) -> Self {
+            RetryDelay {
+                delay: delay_for(Duration::from_millis(0)),
+                attempt: 0,
+                max_attempts,
+            }
+        }
 
-            for cap in url_regex.captures_iter(&body) {
-                return Ok(cap[1].to_string());
+        // our retry delay implements an [exponential backoff](TODO), which simply waits twice as
+        // long between each attempt.
```

Comment:
what does this comment mean?

Reply:
is just giving an explanation for why the duration is set the way it is; added a link to wikipedia in case someone is unfamiliar with exponential backoff.

Reply:
"long between each attempt"

Reply:
scroll over? "which simply waits twice as long between each attempt"